target-component-accessor-compiler/ksp/src/main/kotlin/me/tatarka/inject/target/component/accessor/compiler/ksp/TargetComponentAccessorProcessor.kt

```diff
@@ -0,0 +1,146 @@
+            }
+
+            val funcReceiver = symbol.validateReceiver()
+            if (funcReceiver == null) {
+                logger.error("$funcName should be declared as an extension of a @Component's companion")
+                continue
+            }
+
+            val funcReturn = symbol.validateReturnType()
+            if (funcReturn == null) {
+                logger.error("$funcName's return type should be a type annotated with @Component")
+                continue
+            }
+
+            generate(
+                symbol = symbol,
+                name = funcName,
+                receiverType = funcReceiver,
+                returnType = funcReturn,
+                codeGenerator = codeGenerator
+            )
+        }
+
+        return emptyList()
+    }
+}
+
+private fun KSFunctionDeclaration.validateReceiver(): KSTypeReference? {
+    val funcReceiver = extensionReceiver
+    val funcReceiverName = funcReceiver?.resolve()?.declaration?.qualifiedName?.asString()
+    return funcReceiver?.takeIf {
+        funcReceiverName != null && funcReceiverName.endsWith("Companion")
+    }
+}
+
+private fun KSFunctionDeclaration.validateReturnType(): KSDeclaration? {
+    return returnType?.resolve()?.declaration?.takeIf { funcReturn ->
+        funcReturn.toString() != "Unit" &&
+            funcReturn.annotations.any { it.shortName.asString() == "Component" }
+    }
+}
+
+private fun generate(
+    symbol: KSFunctionDeclaration,
+    name: String,
+    receiverType: KSTypeReference,
+    returnType: KSDeclaration,
+    codeGenerator: CodeGenerator
+) {
+    val file = FileSpec.builder(
+        packageName = symbol.packageName.asString(),
+        fileName = "ActualInject${returnType.simpleName.asString()}",
+    ).apply {
+        addFunction(
+            FunSpec.builder(name).apply {
+                addAnnotation(TargetComponentAccessor::class)
+
+                addModifiers(
+                    symbol.getVisibility().toKModifier() ?: KModifier.INTERNAL,
+                    KModifier.ACTUAL,
+                )
+
+                receiver(receiverType.toTypeName())
+
+                for (param in symbol.parameters) {
+                    val paramName = requireNotNull(param.name?.asString()) {
+                        "$name cannot have a parameter without a name"
+                    }
+
+                    addParameter(paramName, param.type.toTypeName())
+                }
+
+                val funcParams = symbol.parameters.joinToString { it.name?.asString().toString() }
+
+                addCode(
+                    CodeBlock.of(
+                        "return %T.create($funcParams)",
+                        receiverType,
+                    ),
+                )
+
+                // if we're at this point then returnType can't be null
+                returns(symbol.returnType!!.toTypeName())
+            }.build(),
+        )
+    }.build()
+
+    file.writeTo(codeGenerator, aggregating = true)
```

Comment:
why is this aggregating, there's a 1:1 relationship between source and generated no?

Reply:
Just copy paste :smile: